nixos/hypridle: only run the service on Hyprland

JohnRTitor opened the pull request:
Since https://github.com/NixOS/nixpkgs/pull/355959 have been merged we can finally get this going.

This PR makes the service only start with "wayland-session@Hyprland.target". This needs a UWSM configured setup.

Finally closes https://github.com/NixOS/nixpkgs/issues/347651

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
- [x] Tested, as applicable:
  - With my own configuration

fufexan reviewed (commented):
I don't think this is a good way to check for Hyprland's existence.
Take the scenario where a user logs into Hyprland, closes it, and then logs into another compositor/DE/WM. The directory will still exist until the next boot (or whenever it is removed), and hypridle will start every time, even on sessions where it may be unwanted.

JohnRTitor commented:
Well, this is kind of safeguards it instead of just letting it start once `graphical-session.target` has reached.

Another approach would be checking for a variable, like `XDG_SESSION_DESKTOP`.
```nix
  systemd.user.services.hypridle.unitConfig.ConditionEnvironment = "XDG_SESSION_DESKTOP=Hyprland";
```

JohnRTitor commented:
CC @Vladimir-csp for inputs, if they have a better idea :D

Vladimir-csp commented:
My idea is to reuse `systemd-xdg-autostart-condition` which is a predictable mechanism from systemd's implementation of XDG Autostart. I would avoid custom solutions like probing for compositor-specific files or content of `XDG_SESSION_DESKTOP`.

JohnRTitor commented:
Not familiar with the implementation. AFAIK you need a desktop file for XDG autostart?

JohnRTitor commented:
> This would work only if `XDG_SESSION_DESKTOP` is set and only if it is set to `Hyprland`. Uwsm does that (it takes the first item of `XDG_CURRENT_DESKTOP`), but other solutions may not. 

 _Originally posted by @Vladimir-csp in [0021223](https://github.com/NixOS/nixpkgs/commit/0021223389162bffd43b58d466a925087181abf4#r149014092)_

I suppose so, but why Hyprland does not set the variable by default (as in force it)?

fufexan commented:
The 2nd option in this comment has the code for XDG autostart condition https://github.com/NixOS/nixpkgs/issues/347651#issuecomment-2465911345.

Vladimir-csp commented:
Pointing `WantedBy` to a specific unit (or units) is also perfectly valid.

JohnRTitor commented:
> I'm all for it. Though I'm unsure of the route we should take to make hypridle depend on the Hyprland session:
> 
> 1. We simply add `.wantedBy = "wayland-session@hyprland\x2duwsm.desktop.target";`
> 2. We add `.serviceConfig.ExecCondition = ''${pkgs.systemd}/lib/systemd/systemd-xdg-activation-condition "Hyprland" ""'';`
> 
> Option 1 would allow users to add other targets as well (to run hypridle with other compositors), while option 2 would limit hypridle's use to Hyprland only.


Option 1 is not feasible without UWSM, we want to have it working without relying on UWSM if possible. A question also arises if Hypridle even works on other compositors.

Option 2 is good, but I suppose this will also work when UWSM is enabled? (Does Hyprland implement XDG-activation-condition by itself?)

JohnRTitor commented:
I'd lean towards the WantedBy solution then. But that just that alone won't solve our issue.

If we have two `wantedBy`s for example (without the conditions set on this PR), "graphical-session.target" and "wayland-session@Hyprland.target" it will still launch on other DEs not solving our issue. Again, we have to leave graphical-session be, as it should work without UWSM.

This will just bring us back to where we started.

Vladimir-csp commented:
But which unit binds to  `graphical-session.target` in alternative solutions?

fufexan commented:
`WantedBy = "graphical-session.target"` has to go. If there's any systemd user not willing to transition to uwsm, they will need to use `exec-once = systemctl --user start hypridle`. This is a breaking change and has been/will be marked as such.

JohnRTitor commented:
> `WantedBy = "graphical-session.target"` has to go

If using UWSM is the only recommended solution for systemd support in Hyprland by upstream, we can remove `graphical-session.target` in its entirety from here.

I think we should start by adding a withUWSM option in Hyprland module, to make it easier for users and enable it by default later. I'll open a PR soon.

JohnRTitor committed:
nixos/hyprland: only run the service on Hyprland

This needs uwsm enabled. You can do so by setting `programs.hyprland.withUWSM = true`

JohnRTitor reviewed (commented):
![image](https://github.com/user-attachments/assets/753dde9a-0286-4e75-ade9-8be8e9f6d5f3)

nixos manual built

fufexan reviewed (approved):
I think we should provide some information about setting `systemd.user.services.hypridle.wantedBy` to other values to allow using hypridle with other compositors.

I'd prefer this over duplicating the `wantedBy` option.

fufexan commented:
Also, the target will be different when using UWSM. In my case it is `wayland-session@hyprland\x2duwsm.desktop.target`. I assume it will be the same for everyone using the `withUWSM` option, due to how systemd escapes the desktop entry's path (`hyprland-uwsm.desktop`).

JohnRTitor commented:
Sure, but atleast one wantedBy here is needed, else NixOS module system won't recognise the service AFAIK.

JohnRTitor commented:
> Also, the target will be different when using UWSM.

Is that supposed to work like that? 

![image](https://github.com/user-attachments/assets/49e26e4d-35c4-4733-aaa7-da998454cfc8)

I am using the standard configuration of Hyprland with UWSM.

https://github.com/JohnRTitor/nix-conf/blob/6b71ccfb2a1cc273a8c37bb747f736b9f3c1690c/system/hyprland/session.nix#L21-L27

fufexan commented:
Ok, this might be GDM at play here. I'm using greetd's initial_session (autologin) with `uwsm start hyprland-uwsm.desktop`. I thought it would process the file and get the required info from it, instead of passing the name verbatim.

@Vladimir-csp could this be implemented? Or should I modify my call?

Vladimir-csp commented:
Unit specifier comes from whatever was given to `uwsm start` as the main argument, that is how uwsm in the units knows what to search for and do stuff with. In case of `uwsm start hyprland-uwsm.desktop` the resulting specifier will indeed be `hyprland\x2duwsm.desktop`. No double wrapping will happen, since uwsm is capable of parsing `Exec=uwsm start ...` from such entries. But launching non-uwsm entries by uwsm is preferable. As is launching uwsm entries via DMs.

JohnRTitor commented:
> resulting specifier will indeed be `hyprland\x2duwsm.desktop`

perhaps, in case of Hyprland or Sway, autodetection could be done and the service will always be: `wayland-session@Hyprland.target` or `wayland-session@Sway.target`, regardless of how it is launched?

fufexan commented:
Hardcoding that would be ugly. Treating the `Exec` line inside the desktop file (if it has `uwsm start`) verbatim would be a way to go. Though this could be abused to produce an infinite loop.

Atemu commented on nixos/modules/services/wayland/hypridle.nix:
```diff
@@ -10,17 +10,31 @@ let
+    systemd.packages = [ cfg.package ];
+
+    systemd.user.services.hypridle = {
+      # Service should be only be started on Hyprland
+      # this target is started by UWSM
+      wantedBy = [ "wayland-session@Hyprland.target" ];
```
This will only start the service. It will not wait for to have started and never stop it either.

Atemu reviewed (changes requested):
This does not work. At the minimum, you need what I worked out here:

https://github.com/NixOS/nixpkgs/issues/347651#issuecomment-2465928665

Vladimir-csp commented on nixos/modules/services/wayland/hypridle.nix:
```diff
@@ -10,17 +10,31 @@ let
+    systemd.user.services.hypridle = {
+      # Service should be only be started on Hyprland
+      # this target is started by UWSM
+      wantedBy = [ "wayland-session@Hyprland.target" ];
```
```suggestion
      # start only with Hyprland units
      wantedBy = [
        "wayland-session@Hyprland.target"
        "wayland-session@hyprland.desktop.target"
        "wayland-session@hyprland\x2duwsm.target"
      ];
      # order start after compositor (and WAYLAND_DISPLAY) is available
      # order stop before compositor
      after = [ "graphical-session.target" ];
      # stop with the session
      partOf = [ "graphical-session.target" ];
```

Atemu replied:
These two targets do not exist on NixOS. We've explicitly limited this module to only work with the `programs.hyprland.withUWSM` option which only creates the `wayland-session@Hyprland.target`. (I'd prefer a stronger API than the Hyprland package's binary name here but that's something for our uwsm module to offer.

We also want to order hypridle `before` the wayland session so that it's only considered up when hypridle has started up.

Vladimir-csp replied:
edited

Atemu replied:
In my config I made it `before` and `partOf` `wayland-session@Hyprland.target` because I think of hypridle as belonging to the "Hyprland session" rather than the graphical-session directly. This would also be okay IMV.

I do find it a bit weird to depend on the session target rather than the wm service though as that's our actual dependency.

Vladimir-csp replied:
Units of service and target are bound, so there is no difference for dependencies. Targets are more frequently used as dependencies.
```
$ grep -r ^WantedBy= /lib/systemd/system | grep -cF .target
124
$ grep -r ^WantedBy= /lib/systemd/system | grep -cF .service
90
$ grep -r ^WantedBy= /lib/systemd/user | grep -cF .target
24
$ grep -r ^WantedBy= /lib/systemd/user | grep -cF .service
17
```

Vladimir-csp replied:
`wayland-wm@.service` is bound to `wayland-session@.target`, so they are interchangeable as dependencies (but can differ when ordering).

Targets are more frequently used for dependencies than services.

JohnRTitor replied:
Depending on `graphical-session.target`, is fine in my opinion.

When a WM managed by UWSM is terminated, graphical-session also stops. So I understand the partOf bit.

What I don't understand is why are we supposed to start it "before" `graphical-session`? Starting it after isn't correct?

> I'd prefer a stronger API than the Hyprland package's binary name here

Agreed.

Atemu commented on nixos/modules/services/wayland/hypridle.nix:
```diff
@@ -10,17 +10,31 @@ let
+        Whether to enable Hypridle, Hyprland's idle daemon.
+
+        ::: {.note}
+        For this service to work properly on Hyprland, you
+        need to have `programs.hyprland.withUWSM` enabled.
```
We should assert that.

JohnRTitor replied:
No, because @fufexan wants it to be compatible with other compositors.

Atemu replied:
As it stands right now, it cannot be compatible with other compositors because it depends on the hyprland-uwsm-specific units.

If you wanted to use it with another compositor, you should bring your own hypridle unit.

Vladimir-csp replied:
It could be compatible if you just add more targets in `WantedBy` and `After`.

fufexan replied:
Or simply add `systemd.user.services.swayidle.wantedBy = ["your-compositors-session.target"];`. This should be an admonition in the enable option's description.

JohnRTitor commented:
@fufexan just so we are on the clear here, is hypridle supported on WMs like sway, wayfire, mir?
